source/http/HttpConnectionManager.cpp

```diff
@@ -0,0 +1,294 @@
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <algorithm>
+#include <aws/crt/http/HttpConnectionManager.h>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        namespace Http
+        {
+            HttpClientConnectionManagerOptions::HttpClientConnectionManagerOptions()
+                : bootstrap(nullptr), initialWindowSize(SIZE_MAX), port(0), max_connections(2)
+            {
+                AWS_ZERO_STRUCT(socketOptions);
+                AWS_ZERO_STRUCT(hostName);
+            }
+
+            HttpClientConnectionManager::HttpClientConnectionManager(
+                const HttpClientConnectionManagerOptions &connectionManagerOptions,
+                Allocator *allocator) noexcept
+                : m_allocator(allocator), m_good(true), m_lastError(AWS_ERROR_SUCCESS),
+                  m_outstandingVendedConnections(0), m_pendingConnections(0)
+            {
+                m_initialWindowSize = connectionManagerOptions.initialWindowSize;
+                m_bootstrap = connectionManagerOptions.bootstrap;
+                assert(connectionManagerOptions.hostName.ptr && connectionManagerOptions.hostName.len);
+                m_hostName =
+                    String((const char *)connectionManagerOptions.hostName.ptr, connectionManagerOptions.hostName.len);
+                m_max_size = connectionManagerOptions.max_connections;
+                m_port = connectionManagerOptions.port;
+                m_socketOptions = *connectionManagerOptions.socketOptions;
+
+                if (connectionManagerOptions.tlsConnectionOptions)
+                {
+                    m_tlsConnOptions = *connectionManagerOptions.tlsConnectionOptions;
+
+                    if (!m_tlsConnOptions)
+                    {
+                        m_lastError = aws_last_error();
+                        m_good = false;
+                    }
+                }
+            }
+
+            HttpClientConnectionManager::~HttpClientConnectionManager()
+            {
+                Vector<std::shared_ptr<HttpClientConnection>> connectionsCopy = m_connections;
+                /* make sure all connections we know about are closed. */
+                for (auto &connection : connectionsCopy)
+                {
+                    connection->Close();
+                }
+
+                {
+                    /* wait for the last one to shutdown so the memory will be cleaned up. */
+                    std::unique_lock<std::mutex> m_semaphoreLock(m_connectionsLock);
+                    m_shutdownSemaphore.wait(m_semaphoreLock, [this] { return m_connections.empty(); });
```

Comment:
This can deadlock if the ConnectionMgr is released from an event-loop thread. And we want users to set their apps up as event-driven, so that's pretty likely isn't it?

Reply:
wasn't needed anyways.